PRESIDECMS-2214 batch all matching current filter

DominicWatson opened the pull request:
Adding to the work done to make batch edit and delete work in a background task - we have "select all records matching current filter". Merging this back into the background task branch as this is merged from there.

DominicWatson committed:
[twgit] Init feature 'feature-PRESIDECMS-2214_batch-all-matching-current-filter'.

DominicWatson committed:
Merge branch 'feature-PRESIDECMS-2213_batch-edit-delete-background-task' into feature-PRESIDECMS-2214_batch-all-matching-current-filter

DominicWatson committed:
PRESIDECMS-2214 HTML and javascript behaviour for showing option to select all filtered records when doing batch operations

Still to do: actually submit the filter + use that in batch operations

DominicWatson committed:
PRESIDECMS-2214 frontend logic to prepare batch action form with filter data when selecting all records

Stil TODO: actual server side logic to make this work :) :) :)

DominicWatson committed:
PRESIDECMS-2214 return args used to fetch data when fetching data for grid tables

DominicWatson committed:
PRESIDECMS-2214 store hash of source string to prevent tampering with it later (we will use this to validate the input when using for batch operations)

DominicWatson committed:
PRESIDECMS-2214 rename variable inline with naming convention of other returned data

DominicWatson committed:
PRESIDECMS-2214 refactor frontend implementation to use returned data source string for batch operation submission

DominicWatson committed:
PRESIDECMS-2214 validate batch source in multi action handler

DominicWatson committed:
PRESIDECMS-2214 allow adhoc tasks to be interruptable with the existing toothless cancel task function

DominicWatson committed:
2214 implement batch all delete

DominicWatson committed:
PRESIDECMS-2214 simplify logging

and fix little logical bug with do/while loop clause

DominicWatson committed:
PRESIDECMS-2214 Implement batch edit using 'select all records' feature

DominicWatson committed:
PRESIDECMS-2214 rename batch_edit_queue to batch_operation_queue so as to be more useful

DominicWatson committed:
PRESIDECMS-2214 provide an API into operation queue

DominicWatson committed:
PRESIDECMS-2214 add bg task to cleanup old queued batch operations

DominicWatson committed:
PRESIDECMS-2214 fix missing elvis for uncertain url param

DominicWatson committed:
PRESIDECMS-2214 separate out service layer for batch operations into its own service

DominicWatson committed:
PRESIDECMS-2214 documentation tweaks

DominicWatson committed:
PRESIDECMS-2214 make batch delete use the operation queue rather than evaluating filter on each loop (avoids potential for endless loop)

DominicWatson committed:
PRESIDECMS-2214 normalize variable names

DominicWatson committed:
PRESIDECMS-2214 update i18n to indicate that batch operation will apply to roughly the number of records shown

DominicWatson committed:
PRESIDECMS-2214 remove unwanted variable

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/pixl8/Preside-CMS?pullRequest=1038) <br/>All committers have signed the CLA.

DominicWatson committed:
PRESIDECMS-2214 tweaks to make batch delete operation have its own pre/post hooks that code should update to (doesn't make sense to use the old hooks)

Also optimize the queue fetching so that we delete items from the queue as soon as we fetch them (by default)

DominicWatson committed:
PRESIDECMS-2214 fix and add test

Pixl8RusselCole commented on system/services/admin/DataManagerBatchOperationService.cfc:
```diff
@@ -0,0 +1,429 @@
+		try {
+			do {
+				if ( arguments.batchAll ) {
+					sourceIds = getNextBatchRecordsFromQueue( queueId, 100 );
```
Slightly concerns me having 100 hard-coded in 2 places. Potentially someone could change the top one to 1000 without knowing to change the other

Pixl8RusselCole commented on system/services/admin/DataManagerBatchOperationService.cfc:
```diff
@@ -0,0 +1,429 @@
+		return queueId;
+	}
+
+	public numeric function clearBatchOperationQueue( required string queueId ) {
```
I wonder if there should be a message here to say how many records were removed from queue?

Pixl8RusselCole commented on system/services/admin/DataManagerBatchOperationService.cfc:
```diff
@@ -0,0 +1,429 @@
+			clearBatchOperationQueue( queueId );
+		}
+		if ( canInfo ) {
+			arguments.logger.info( $translateResource( uri="cms:datamanager.batchedit.task.finished.message", data=[ objectTitle, fieldTitle, NumberFormat( totalRecords ) ] ) );
```
Is totalRecords right in the finished.message? The task could have been interrupted before that number of records has been processed

DominicWatson committed:
PRESIDECMS-2214 just use default pagination argument in our calls

DominicWatson committed:
PRESIDECMS-2214 show numbers of cancelled queued operations when cancelling out of a task early

DominicWatson committed:
PRESIDECMS-2214 show actual processed count for final log message on record updates

DominicWatson committed:
PRESIDECMS-2214 ensure that processed variable is incremented no matter what scenario we're in
